README.md

```diff
@@ -350,14 +367,22 @@ The only reserved ability MUST be the un-namespaced [`"*"` or "superuser"](#41-s
 
 ### 3.2.5 Proof of Delegation
 
-The `prf` field MUST contain UCAN witnesses (the "inputs" of a UCAN). As they need to be independently verifiable, proofs MUST either be the fully encoded version of a UCAN (including the signature) or the content address of the relevant proof. Attenuations not covered by a proof in the `prf` array MUST be treated as owned by the issuer DID.
+The `prf` field MUST contain the content address <!-- FIXME add link to CID discussion later in doc? --> of UCAN proofs (the "inputs" of a UCAN). Attenuations not covered by a proof in the `prf` array MUST be treated as <!-- FIXME as invalid, because we now have my: and as: --> owned by the issuer DID.
 
-Proofs referenced by content address MUST be resolvable by the recipient, for instance, over a DHT or database. Since they have no availability concern, inlined proofs SHOULD be preferred unless the proof chain has become too large for a request.
+Proofs referenced by content address MUST be resolvable by the recipient, for instance, over a DHT or database. <!-- FIXME reference (or move to) lookup section -->
 
 #### Examples
 
 ``` json
-"prf": ["eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsInVhdiI6IjAuMS4wIn0.eyJhdWQiOiJkaWQ6a2V5OnpTdEVacHpTTXRUdDlrMnZzemd2Q3dGNGZMUVFTeUExNVc1QVE0ejNBUjZCeDRlRko1Y3JKRmJ1R3hLbWJtYTQiLCJpc3MiOiJkaWQ6a2V5Ono1QzRmdVAyRERKQ2hoTUJDd0FrcFlVTXVKWmROV1dINU5lWWpVeVk4YnRZZnpEaDNhSHdUNXBpY0hyOVR0anEiLCJuYmYiOjE1ODg3MTM2MjIsImV4cCI6MTU4OTAwMDAwMCwic2NwIjoiLyIsInB0YyI6IkFQUEVORCIsInByZiI6bnVsbH0.Ay8C5ajYWHxtD8y0msla5IJ8VFffTHgVq448Hlr818JtNaTUzNIwFiuutEMECGTy69hV9Xu9bxGxTe0TpC7AzV34p0wSFax075mC3w9JYB8yqck_MEBg_dZ1xlJCfDve60AHseKPtbr2emp6hZVfTpQGZzusstimAxyYPrQUWv9wqTFmin0Ls-loAWamleUZoE1Tarlp_0h9SeV614RfRTC0e3x_VP9Ra_84JhJHZ7kiLf44TnyPl_9AbzuMdDwCvu-zXjd_jMlDyYcuwamJ15XqrgykLOm0WTREgr_sNLVciXBXd6EQ-Zh2L7hd38noJm1P_MIr9_EDRWAhoRLXPQ"]
+"prf": ["bafkreihogico5an3e2xy3fykalfwxxry7itbhfcgq6f47sif6d7w6uk2ze"]
+
+```
+
+Which resolves as <!-- come back and flesh out -->
+```json
+{
+  "bafkreihogico5an3e2xy3fykalfwxxry7itbhfcgq6f47sif6d7w6uk2ze": "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsInVhdiI6IjAuMS4wIn0.eyJhdWQiOiJkaWQ6a2V5OnpTdEVacHpTTXRUdDlrMnZzemd2Q3dGNGZMUVFTeUExNVc1QVE0ejNBUjZCeDRlRko1Y3JKRmJ1R3hLbWJtYTQiLCJpc3MiOiJkaWQ6a2V5Ono1QzRmdVAyRERKQ2hoTUJDd0FrcFlVTXVKWmROV1dINU5lWWpVeVk4YnRZZnpEaDNhSHdUNXBpY0hyOVR0anEiLCJuYmYiOjE1ODg3MTM2MjIsImV4cCI6MTU4OTAwMDAwMCwic2NwIjoiLyIsInB0YyI6IkFQUEVORCIsInByZiI6bnVsbH0.Ay8C5ajYWHxtD8y0msla5IJ8VFffTHgVq448Hlr818JtNaTUzNIwFiuutEMECGTy69hV9Xu9bxGxTe0TpC7AzV34p0wSFax075mC3w9JYB8yqck_MEBg_dZ1xlJCfDve60AHseKPtbr2emp6hZVfTpQGZzusstimAxyYPrQUWv9wqTFmin0Ls-loAWamleUZoE1Tarlp_0h9SeV614RfRTC0e3x_VP9Ra_84JhJHZ7kiLf44TnyPl_9AbzuMdDwCvu-zXjd_jMlDyYcuwamJ15XqrgykLOm0WTREgr_sNLVciXBXd6EQ-Zh2L7hd38noJm1P_MIr9_EDRWAhoRLXPQ"
+}
```

Comment:
JWTs are base64 encoded. We can explore defining a custom data format, but a JWT is base64

Reply:
Sure but IPLD block does not needs to be bytes of JWT it can be anything from what exactly right JWT can be computed, that is to say it doesn't need to store base64 artifacts those can be computed as IPLD block is serialized to JWT

Reply:
Ah, I would be missing something! My understanding is that IPLD won't be able to capture differences in JWT implementations, so would need to keep the exact signed bytes. Is that not the case?

Reply:
It's also possible to move this out of the format spec, but we do need to have _some_ way of talking about proofs. If we are going to make them always CIDs, there has to be some consistent format to resolve them, no? We can leave that up to various implementations, but that may be a high implementer burden having an n-way codec

Reply:
> Ah, I would be missing something! My understanding is that IPLD won't be able to capture differences in JWT implementations, so would need to keep the exact signed bytes. Is that not the case?

Not exactly IPLD code is essentially pair of `encode(data:T): Bytes` and `decode(bytes:Bytes):T` so as long as you can do `decode(encode(t)) === t` actual byte representation is a codec implementation detail.

With that said dag-ucan can simply store JSON representations without base64 encoding them, it can base64 encode them on decode instead.

Reply:
I see, gotcha. What if you have a proof that was not generated from that library? It may not be guaranteed to pass this check `decode(encode(t)) === t`. I had a bunch of tests that looked like that in my Haskell implementation, but would fail on tokens generated by Node. It's possible that different browsers could have different serialization behaviour, too.

Don't get me wrong, nothing would make me happier than a dag-ucan! I'm trying to balance this against the ubiquity and marketing familiarity of JWT.

Reply:
For what it's worth dag-ucan implementation I wrote does something slightly different:

```ts
encode(data:UCAN): Bytes
decode(data:Bytes): UCAN
format(data:UCAN): JWT
parse(jwt:JWT): UCAN
```

Where `UCAN` is kind of IR representation like:

```ts
{
   // Bytes for `JSON.stringify(head)`
   header: Bytes,
   // Bytes for `JSON.stingify(payload)`
   body: Bytes,
   // Bytes for actual signature
   signature: Bytes
}
```

Reply:
Oooooh I see! Indeed, this mirrors what my implementation does during validation! Okay, I think I see where you're going with this now.

Okay, so may be over-specified as one big string, because it forces you to break it at the `.`s yourself during deserialization. 👍 Agreed

Reply:
Are you suggesting that we make the collection type respect this decomposed version? I'm amenable to that!

Reply:
I'm just saying that spec should probably specify what that CID should decode to. And maybe actual JWT string is not the best option.

Reply:
For what it's worth I have shifted direction of dag-ucan implementation to support dual representation. Primary one is fully in dag-cbor from which valid JWT string can be derived. But since not as all UCANs can be represented in primary representation (given that CBOR will disregard whitespaces and force alphabetic order of keys, signatures would not match) secondary fallback representation in RAW codec is used.

I think it might be a good idea to support CIDs in those two formats:

1. [Raw code 0x55](https://github.com/multiformats/multicodec/blob/master/table.csv#L39)
2. UCAN code 0x78c0 (I have not submitted request to add it to multicodec table yet)

I do not think it is worth supporting DAG-PB encoding because only advantage it has over RAW is in representing files that don't fit libp2p block size limit of ~1MiB, and probably it is bad idea to have such a large UCAN anyway.

Reply:
Is the raw codec (`0x55`) in there to satisfy the inlining option?

> https://github.com/multiformats/multicodec/pull/264

Oh fabulous; this was on my backlog 🙌 I'll update this spec to reflect `0x78c0 👍

(btw: is there some meaning behind `0x78C0`?)

> given that CBOR will disregard whitespaces and force alphabetic order of keys, signatures would not match

My understanding is that not all CBOR encoders will do this, only the ones that elect for deterministic encoding. It may be worth specifying that where it's used. There are use cases for forcing a non-alphanumeric order in some protocols to make them more wire-friendly (`b5`, Philipp, myself, and others were exploring this as part of a WNFS-in-IPFS experimentation). [From the CBOR spec](https://www.rfc-editor.org/rfc/rfc8949.html#name-deterministically-encoded-c):

> Some protocols may want encoders to only emit CBOR in a particular deterministic format; those protocols might also have the decoders check that their input is in that deterministic format. Those protocols are free to define what they mean by a "deterministic format" and what encoders and decoders are expected to do. This section defines a set of restrictions that can serve as the base of such a deterministic format.

Reply:
> Is the raw codec (0x55) in there to satisfy the inlining option?

No it's there to allow representing UCANs that can be correctly represented in CBOR encoding.

> Oh fabulous; this was on my backlog 🙌 I'll update this spec to reflect 0x78c0 👍
>
> (btw: is there some meaning behind 0x78C0?)

Please note that code may still change. No real meaning other than I tired firs couple of letters to look different from the common CIDs.

> My understanding is that not all CBOR encoders will do this, only the ones that elect for deterministic encoding. I guess you mean dag-cbor specifically?

Yes I was referring to dag-cbor. By encoder I meant IPLD multicodec, should have been more specific.

> It may be worth specifying that where it's used. 

Specifying where ? I'm not sure I follow this

Reply:
> Specifying where ? I'm not sure I follow this

Sorry, actually this is entirely my bad! This is for dag-ucan, not UCAN in general 👍 I had meant to specify it in dag-ucan, but that is already the case.

Reply:
> I don't see the codec in there yet.

This is what we call codecs in JS IPLD land https://github.com/multiformats/js-multiformats/blob/9bcd7fef62888d7cefe8e4f5e929d4e3c9dadda9/src/codecs/interface.ts#L4-L21

Per that definition module is a codec as it exports `code`, `encode`, `decode`.

I'll comment in the https://github.com/ipld/js-dag-ucan/pull/6 so it' easier to point to specific things

Reply:
> Hmm I think my brain just caught up. 0x780C would require the dag-ucan decoder specifically, which uses an internal deterministic CBOR representation. My understanding is that the actual signature is against the base64url JWT as normal.

That is accurate. That is also how we can format such UCANs into JWTs

Reply:
> someone resolving this CID will get back CBOR, yes?

I'm not sure what you mean exactly, but yeah if you were to try and get DAG for such CID it would require dag-ucan codec which internally utilized dag-cbor.